include/ginkgo/core/solver/workspace.hpp

```diff
@@ -0,0 +1,204 @@
+SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+(INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+******************************<GINKGO LICENSE>*******************************/
+
+#ifndef GKO_PUBLIC_CORE_SOLVER_WORKSPACE_HPP_
+#define GKO_PUBLIC_CORE_SOLVER_WORKSPACE_HPP_
+
+
+#include <typeinfo>
+
+
+#include <ginkgo/core/matrix/dense.hpp>
+
+
+namespace gko {
+namespace solver {
+namespace detail {
+
+
+/**
+ * Type-erased object storing any kind of gko::array
+ */
+class any_array {
+public:
+    template <typename ValueType>
+    array<ValueType>& init(std::shared_ptr<const Executor> exec, size_type size)
+    {
+        auto container = std::make_unique<concrete_container<ValueType>>(
+            std::move(exec), size);
+        auto& arr = container->arr;
+        data_ = std::move(container);
+        return arr;
+    }
+
+    bool empty() const { return data_.get() == nullptr; }
+
+    template <typename ValueType>
+    bool contains() const
+    {
+        return dynamic_cast<const concrete_container<ValueType>*>(data_.get());
+    }
+
+    template <typename ValueType>
+    array<ValueType>& get()
+    {
+        GKO_ASSERT(this->template contains<ValueType>());
+        return dynamic_cast<concrete_container<ValueType>*>(data_.get())->arr;
+    }
+
+    template <typename ValueType>
+    const array<ValueType>& get() const
+    {
+        GKO_ASSERT(this->template contains<ValueType>());
+        return dynamic_cast<const concrete_container<ValueType>*>(data_.get())
+            ->arr;
+    }
+
+    void clear() { data_.reset(); }
+
+private:
+    struct generic_container {
+        virtual ~generic_container() = default;
+    };
+
+    template <typename ValueType>
+    struct concrete_container : generic_container {
+        template <typename... Args>
+        concrete_container(Args&&... args) : arr{std::forward<Args>(args)...}
+        {}
+
+        array<ValueType> arr;
+    };
+
+    std::unique_ptr<generic_container> data_;
+};
+
+
+class workspace {
+public:
+    workspace(std::shared_ptr<const Executor> exec) : exec_{std::move(exec)} {}
+
+    workspace(const workspace& other) : workspace{other.get_executor()} {}
+
+    workspace(workspace&& other) : workspace{other.get_executor()}
+    {
+        other.clear();
+    }
+
+    workspace& operator=(const workspace& other) { return *this; }
+
+    workspace& operator=(workspace&& other)
+    {
+        other.clear();
+        return *this;
+    }
+
+    template <typename VectorType, typename CreateOperation>
+    VectorType* create_or_get_vector(int vector_id, CreateOperation op,
+                                     const std::type_info& expected_type,
```

Comment:
using a Solver with both a custom Dense subtype and a regular type, not sure if that is realistic, I would like it to work correctly with create_with_config_of and create_with_type_of.

Reply:
So first, this affects none of our current solvers, since there we don't use the `create_or_get_vector` functions directly, right? And also, no custom solver that derives from `SolverBase` would be affected. 
As I understand it now, the only situation where what you mentioned makes sense is something like:
```
Dense* p = ws.create_or_get_vector<Dense>(..., typeid(DenseSubtype), ...)
```
But I'm still not sure if that is what you actually mean, since in that case you could just do
```
Dense* p = ws.create_or_get_vector<DenseSubtype>(...,  typeid(DenseSubtype), ...)
```
Honestly, a construct that uses `expected_type != typeid(VectorType{})` seems to be quite hard to understand.

Reply:
in MFEM's case, you can both use a ginkgo matrix in an MFEM solver as well as an MFEM operator in a ginkgo solver. This means that we need to support the case that the MFEM "system matrix" inside a Ginkgo solver needs a specific (i.e. MFEM) vector type.

Reply:
Since the workspace stores the vectors as LinOp ptrs, I guess you mean that the workspace should also support types that are not derived from Dense (or whatever we will use as a multivector base type in the future)? If that is the case, then your approach makes sense, but I would like to think over that assumption for a bit.

Reply:
yes, supporting both Library Operator applied to Ginkgo Dense and Ginkgo LinOp applied to Library Vector is a bit of a mixed bag, but that's what was required for the MFEM integration to work, is that right @nbeams? Or could we do away with one of the two directions?